### PR TITLE
120 rejection handler

### LINKED ISSUE #120: Make handleOAuthRequestMiddleware configurable in case of rejection
Currently, if `handleOAuthRequestMiddleware` fails to validate the token of a request then it automatically rejects the request, see https://github.com/zalando-incubator/authmosphere/blob/master/src/express-tooling.ts#L108 and https://github.com/zalando-incubator/authmosphere/blob/master/src/express-tooling.ts#L114

`rejectRequest` (https://github.com/zalando-incubator/authmosphere/blob/master/src/utils.ts#L125) (or better the behaviour of it), should be configurable since it is now quite intransparent how the flow works.

### PR DESCRIPTION
**Pull request by ISO50**
close #120

### CODE REVIEW

File: src/express-tooling.ts [hunk 1]
```diff
@@ -28,82 +29,102 @@ import { safeLogger } from './safe-logger';
 const AUTHORIZATION_HEADER_FIELD_NAME = 'authorization';
 
 /**
- * Returns a function (express middleware) that validates the scopes against the user scopes
- * attached to the request (for example by `handleOAuthRequestMiddleware`).
- * If the the requested scopes are not matched request is rejected (with 403 Forbidden).
+ * Returns a function (express middleware) that validates the scopes against the user scopes.
```
Comment (bzums) [inline src/express-tooling.ts]: "[...]validates the scopes against the user scopes"
Just some wording: what exactly are "the scopes" and "the user scopes" and where do they come from?
Suggestion:
"[...]validates the scopes attached to the request against the specified required scopes"

File: src/express-tooling.ts [hunk 2]
```diff
@@ -28,82 +29,102 @@ import { safeLogger } from './safe-logger';
...
+        logOrNothing.warn(`Error while executing precedenceFunction: ${error}`);
+        // PrecedencFunction was not successful
+        //  false -> trigger fallback to default scope validation
+        return false;
+      })
       .then(isAllowed => {
-        if (isAllowed) {
-          next();
+        if (isAllowed) { // precedenceFunction grants access
+          logOrNothing.debug('Precedence function grants access.');
```
Comment (bzums) [inline src/express-tooling.ts]: Remove "." for consistency 👮

Comment (bzums) [reply src/express-tooling.ts]: and maybe adapt to: "precedenceFunction grants access"? Like line 82

File: test/unit/express-tooling/authentication-middleware.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        ...createRequestMock([])
+      } as any as Request;
+
+      // when
+      authenticationMiddleware(config)(requestMock, createResponseMock(), next);
+
+      // then
+      setTimeout(() => {
+        // tslint:disable-next-line
+        expect(next).to.have.been.called;
```
Comment (bzums) [inline test/unit/express-tooling/authentication-middleware.ts]: `expect(next).to.have.been.calledWith()` would be an alternative where do not have to disable tslint

Comment (bzums) [reply test/unit/express-tooling/authentication-middleware.ts]: sadly that does not work for `not.have.been.called`...

File: test/unit/express-tooling/authentication-middleware.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+      // then
+      setTimeout(() => {
+        // tslint:disable-next-line
+        expect(next).to.not.have.been.called;
+        done();
+      });
+    });
+
+    it('should call custom notAuthenticated handler', (done) => {
```
Comment (bzums) [inline test/unit/express-tooling/authentication-middleware.ts]: "should call custom onNotAuthenticatedHandler" 👮

File: test/unit/express-tooling/require-scopes-middleware.ts [hunk 5]
```diff
@@ -114,15 +115,19 @@ describe('express tooling', () => {
       const requestMock = createRequestMock(['uid']);
       const requiredScopes = ['test'];
 
-      const preOptions = {
+      const precedenceOptions = {
         precedenceFunction: () => {
           return Promise.resolve(true);
         },
         precedenceErrorHandler: () => { return; }
```
Comment (bzums) [inline test/unit/express-tooling/require-scopes-middleware.ts]: remove `precedenceErrorHandler`

File: test/unit/express-tooling/require-scopes-middleware.ts [hunk 6]
```diff
@@ -140,15 +145,19 @@ describe('express tooling', () => {
       const requestMock = createRequestMock(['uid']);
       const requiredScopes = ['test'];
 
-      const preOptions = {
+      const precedenceOptions = {
         precedenceFunction: () => {
           return Promise.resolve(false);
         },
         precedenceErrorHandler: () => { return; }
```
Comment (bzums) [inline test/unit/express-tooling/require-scopes-middleware.ts]: remove precedenceErrorHandler

File: test/unit/express-tooling/require-scopes-middleware.ts [hunk 7]
```diff
@@ -191,15 +205,20 @@ describe('express tooling', () => {
       const requestMock = createRequestMock(['test']);
       const requiredScopes = ['test'];
 
-      const preOptions = {
+      const precedenceOptions = {
         precedenceFunction: () => {
           return Promise.resolve(false);
         },
         precedenceErrorHandler: () => { return; }
```
Comment (bzums) [inline test/unit/express-tooling/require-scopes-middleware.ts]: remove precedenceErrorHandler

File: test/unit/express-tooling/authentication-middleware.ts (hunk 3 above)
Comment (ISO50) [reply test/unit/express-tooling/authentication-middleware.ts]: @bzums 

Discussion @chai issue tracker: https://github.com/chaijs/chai/issues/41
Why chai is crap 😉 : https://github.com/moll/js-must#asserting-on-property-access
TSLINT plugin (eslint already has one): https://github.com/palantir/tslint/issues/2614

In general this happen if you have property accessors with side effects 🤯

File: src/express-tooling.ts [hunk 8]
```diff
@@ -28,82 +29,102 @@ import { safeLogger } from './safe-logger';
...
+        logger,
+        onAuthorizationFailedHandler,
+        precedenceOptions
+      } = options;
+
+      const logOrNothing = safeLogger(logger);
+
+      const authorizationFailedHandler =
+        typeof onAuthorizationFailedHandler === 'function' ?
+          (req: Request, res: Response, next: NextFunction, _scopes: string[], _logger: Logger) =>
```
Comment (Retro64) [inline src/express-tooling.ts]: maybe a type for this signature, as it is used multiple times?

File: test/unit/express-tooling/require-scopes-middleware.ts (hunk 5 above)
Comment (ISO50) [reply test/unit/express-tooling/require-scopes-middleware.ts]: done for whole test file

File: test/unit/express-tooling/require-scopes-middleware.ts (hunk 6 above)
Comment (ISO50) [reply test/unit/express-tooling/require-scopes-middleware.ts]: done for whole test file

File: src/types/AuthenticationMiddlewareOptions.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import { NextFunction, Request, Response } from 'express';
+import { Logger } from '.';
+
+type AuthenticationMiddlewareOptions = {
+  tokenInfoEndpoint: string,
+  logger?: Logger,
+  onNotAuthenticatedHandler?: (request: Request, resonse: Response, next: NextFunction, logger: Logger) => void,
```
Comment (Retro64) [inline src/types/AuthenticationMiddlewareOptions.ts]: as the signature is used several times, might it make sense to define a type?

File: src/express-tooling.ts (hunk 8 above)
Comment (ISO50) [reply src/express-tooling.ts]: Here we could lower the noise a lot with a type.

File: src/types/AuthenticationMiddlewareOptions.ts (hunk 9 above)
Comment (ISO50) [reply src/types/AuthenticationMiddlewareOptions.ts]: Here I don't agree. We do not repeat that type in the middleware implementation, but we could write one and export it for lib consumers.

**Comment by ISO50** [comment]
👍

**Comment by bzums** [comment]
👍

**Comment by ISO50** [comment]
👍

**Pull request merged**